File: app/src/main/java/com/ivanloy/nutricards/ds/Hand.java
package com.ivanloy.nutricards.ds;

public class Hand<T> implements Cloneable{

    private T[] hand;

    public Hand() { hand = null; }

    /**
     * @param cards The initial cards of the hand
     */
    public Hand(T... cards) {

        hand = cards;

    }

    public int size() { return hand == null ? 0 : hand.length; }

    public T[] getHand() { return hand; }

    public void addCard(T... cards) {

        T[] hand_;

        if(hand != null) {

            hand_ = (T[]) new Object[hand.length + cards.length];

            int i;
            for(i = 0; i < hand.length; i++) hand_[i] = hand[i];
            for(; i < hand_.length; i++) hand_[i] = cards[i - hand.length];

        }else {

            hand_ = (T[])new Object[cards.length];
            for(int i = 0; i < cards.length; i++) hand_[i] = cards[i];

        }

        hand = hand_;

    }

    public void addCard(Hand hand) {

        addCard((T[])hand.getHand());

    }

    public Hand clone() {

        return new Hand(hand);

    }

}

File: app/src/main/java/com/ivanloy/nutricards/ds/Hand.kt
package com.ivanloy.nutricards.ds

import com.ivanloy.nutricards.gameelements.FoodCard

class Hand<T>: Cloneable{

    var cardList: MutableList<T> = ArrayList()
        private set

    constructor(hand : Hand<T>){
        this.cardList = hand.cardList.toMutableList();
    }

    constructor(vararg cards : T){
        cardList = cards.toMutableList()
    }

    fun peekCard(index : Int) : T{
        return cardList[index]
    }

    fun peekCardOrDefault(index : Int, default : T) : T{
        return if(index < cardList.size) cardList[index] else default
    }

    fun removeCard(vararg cards: T){
        cards.forEach {
            cardList.remove(it)
        }
    }

    fun removeCardWithPosition(index : Int) : T {
        val card : T = cardList[index]
        cardList.removeAt(index)
        return card
    }

    fun removeCardWithPositionAndReplacement(index : Int, replacement : T) : T {
        val card : T = cardList[index]
        cardList[index] = replacement
        return card
    }

    fun size(): Int {
        return cardList.size
    }

    fun addCard(vararg cards: T) {
        cardList.addAll(cards.toList())
    }

    fun addCard(hand: Hand<T>) {
        cardList.addAll(hand.cardList)
    }

}